src/gradient.toit

```diff
@@ -0,0 +1,303 @@
+      // This is the angle that we will use to calculate the verticalness of the
+      // rectangle.
+      rangle := math.atan (w.to_float / h)  // From 0 to PI/2.
+      rangle *= 180.0 / math.PI            // From 0 to 90.
+      draw_vertical_ = angle < rangle
+      texture_length/int := ?
+      texture_length_repeats_ = 1
+      if draw_vertical_:
+        // The gradient is more vertical than the rectangle, so we will draw
+        // vertical lines on the rectangle.
+        texture_length = (h + w * (math.tan (angle * math.PI / 180.0)) + 0.01).round
+        if angle == 0:
+          // For efficiency, repeat the gradient a few times in the buffer.
+          texture_length = h
+          texture_length_repeats_ = max 1 (512 / texture_length)
+      else:
+        // The gradient is more horizontal than the rectangle, so we will draw
+        // horizontal lines on the rectangle.
+        texture_length = (w + h * (math.tan ((90 - angle) * math.PI / 180.0)) + 0.01).round
+        if angle == 90:
+          // For efficiency, repeat the gradient a few times in the buffer.
+          texture_length = w
+          texture_length_repeats_ = max 1 (512 / texture_length)
+      texture_length_ = texture_length
+
+      red_pixels_ = ByteArray texture_length * texture_length_repeats_
+      green_pixels_ = ByteArray texture_length * texture_length_repeats_
+      blue_pixels_ = ByteArray texture_length * texture_length_repeats_
+      ranges/List := extract_ranges_ specifiers
+      ranges.do: | range |
+        get_colors range texture_length: | index red green blue |
+          red_pixels_[index] = red
+          green_pixels_[index] = green
+          blue_pixels_[index] = blue
+        (texture_length_repeats_ - 1).repeat:
+          red_pixels_.replace ((it + 1) * texture_length) red_pixels_[0..texture_length]
+          green_pixels_.replace ((it + 1) * texture_length) green_pixels_[0..texture_length]
+          blue_pixels_.replace ((it + 1) * texture_length) blue_pixels_[0..texture_length]
+
+  draw canvas/Canvas x/int y/int --autocropped/bool -> none:
+    if not canvas.supports_8_bit: throw "UNSUPPORTED"
+    angle := angle_
+    w := w_
+    h := h_
+    analysis := canvas.bounds_analysis x y w h
+    if analysis == Canvas.DISJOINT: return
+    // Determine whether the draw operations will be automatically cropped for
+    // us, or whether we need to do it ourselves by using slices for drawing
+    // operations.  We could also check whether we are inside a window that will
+    // use compositing to crop everything.
+    if analysis == Canvas.CANVAS_IN_AREA or analysis == Canvas.COINCIDENT: autocropped = true
+
+    // CSS gradient angles are:
+    //    0 bottom to top.
+    //   90 left to right
+    //  180 top to bottom
+    //  270 right to left
+
+    repeats := texture_length_repeats_
+    offset := 0
+    source_width := 0
+    r := null
+    g := null
+    b := null
+
+    draw_block := : | y2 hw lines orientation |
```

Comment:
I think `draw_block` is the wrong name.
It doesn't draw, but updates `r`, `g`, `b`, and `source_width`.

I would probably also pass `offset` in.

Reply:
renamed to `update_r_g_b_block`